lib/compiler/passes/generate-js.js

```diff
@@ -80,9 +85,52 @@ function wrapInSourceNode(prefix, chunk, location, suffix, name) {
   return new SourceNode(null, null, null, [prefix, chunk, suffix]);
 }
 
-// Generates parser JavaScript code.
+/**
+ * @typedef {(string|SourceNode)[]} SourceArray
+ *
+ * @typedef {PEG.SourceBuildOptions<PEG.SourceOutputs>} SourceBuildOptions
+ * @typedef {object} ExtraOptions
+ * @property {undefined} [dependencies]
+ * @property {undefined} [exportVar]
```

Comment:
It's just a typescript trick to make existing code work without adding extra checks. SourceBuildOptions is a union, some elements of which have an optional dependencies, and some don't (similar for exportVar).

But the existing code doesn't care. It says `if (options.dependencies) { ... }`, and if the optional dependencies is there, the clause executes.

But if you type `options` as `SourceBuildOptions`, tsc complains, because `dependencies` might not exist on options. So you either have to add a cast (and I've been trying really hard to avoid those wherever possible - they're ugly), or do some more checking to determine what kind of SourceBuildOptions we've got.

So the trick is to type it as `SourceBuildOptions & ExtraOptions`. This is fine, because anything that can be assigned to `SourceBuildOptions` can be assigned to `SourceBuildOptions & ExtraOptions`, but now I can say `if (options.dependencies) { ... }` again, without typescript complaining.

Basically adding the intersection with ExtraOptions does nothing for the union members where they're already defined and optional (because those fields could already be undefined in those cases), but it allows for them to be set to "undefined" in the other types.

Reply:
TS Playground [link for clarity](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAgjAvDA3jAZge0wLhgZxACcBLAOwHMBuGAIwEMi8yBXAW1qiJoYC8B+Zu05EA2gF0aAXyqhIsAEJJUGbHkKlKPRkI5ca6EgPXFyFCdNlzoMAKrKEAHxgKr4GwFEAHsXrK0fIIwLGQAJlCGZFChBkZBIeGR0TBSwMDoIQDGICSYZBj0JAA2+AAULHi2AJSowDD1MCToMOUAdHw1aHUNPQD0vTAA7gAW9CBQAG5c3fWpqelZOXlDmEQA1mUVdjAAZDDevp0zjc1tHSrHfQMjY5PTPXPAQA)

Reply:
But it isn't quite right... (which showed up when I tried to write a test that calls generateJS directly).

Reply:
It needs to be `unknown`, not `undefined` so that you can call it. See [here](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAgjAvDA3jAZge0wLhgZxACcBLAOwHMBuGAIwEMi8yBXAW1qiJoYC8B+Zu05EA2gF0aAXyqhIsAEJJUGbHkKlKPRkI5ca6EgPXFyFCdNlzoMAKrKEAHxgKr4GwFEAHsXrK0fIIwLGQA1mSYAO5kBkZBIeFRZDBSwMDoIQDGICSYyej0JAA2+AAULHi2AJSowDD1MCToMOUAdHw1aHUNPQD0vTCRABb0IFAAblzd9amp6Vk5eYOYRKFlFXYwAGQw3r6d043NbR0qh30Dw6MTUz2zaRlk2bn52OWVBz0FxetVsj2RFZrcp-YCpIA)

Reply:
Ugh. I give up.

I've definitely used this previously; and in my case it was important, because the members I was adding had different types in different elements of the union. But here, as you say, dependencies is either defined, and of type `Dependencies`, or it's not defined. And exportVar is either a string, or it's not defined. So we can use those types to make this work.

Reply:
Just for my own sanity, the way I've done this in the past (I went and looked) is to add `{ field?: undefined }` to each member of the union that doesn't already have `field`. Blindly adding it to the whole union doesn't work. Adding `{ field?: unknown }` to the whole union /could/ work, but only if you don't care that when you extract the field from the union, it's typed as `unknown` (but if you extract it from a specific member of the union, its still typed correctly).